colossalai/shardformer/layer/_operation.py

```diff
@@ -168,6 +169,133 @@ def backward(ctx, grad_output):
         return grad_input, grad_weight, grad_bias, None, None, None
 
 
+def _ring_as_gather(func, input_to_gather=None, input_local=None, process_group=None, gather_dim=1, keep_item=False):
+    # currently only support one single tensor as output
+    group_size = dist.get_world_size(process_group)
+    cur_rank = dist.get_rank(process_group)
+
+    # output_tensors = [torch.empty((input_shape[0], input_shape[1], weight_shape[0])) for _ in range(group_size)]
+
+    # initialization of ring communication
+    recv_rank = cur_rank + 1 if cur_rank + 1 < group_size else 0
+    send_rank = cur_rank - 1 if cur_rank > 0 else group_size - 1
```

Comment:
They should be global rank. For example, we have 2x2 groups. For process 2, the rank in this group is 0, but we may use 2 rather than 0 here.
See https://github.com/hpcaitech/ColossalAI/blob/4fa689fca1ecf50b8e905cf1c74d4a2c08219daf/colossalai/pipeline/stage_manager.py#L42

Reply:
Yes, thanks for clarification. Corrected.